File: memx_resolver_workflow_plugin/typed_ref.py
"""Strict v2 typed_ref support for the memx-resolver workflow plugin.

The v2 wire format is:
<domain>:<entity_type>:<provider>:<entity_id>

Domain, entity type, and provider are canonicalized to lowercase. Entity IDs
are decoded on input and RFC 3986 encoded on output so colons remain
reversible.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum
from typing import Union
from urllib.parse import quote, unquote_to_bytes


class Domain(str, Enum):
    """Domain represents the system area for typed_ref."""

    MEMX = "memx"
    AGENT_TASKSTATE = "agent-taskstate"
    TRACKER = "tracker"


class EntityType(str, Enum):
    """Compatibility enum for common entity types."""

    EVIDENCE = "evidence"
    EVIDENCE_CHUNK = "evidence_chunk"
    KNOWLEDGE = "knowledge"
    ARTIFACT = "artifact"
    LINEAGE = "lineage"
    DOC = "doc"
    CHUNK = "chunk"
    CARD = "card"
    TASK = "task"


class Provider(str, Enum):
    """Compatibility enum for common providers."""

    LOCAL = "local"
    JIRA = "jira"
    GITHUB = "github"
    LINEAR = "linear"


DEFAULT_PROVIDER = Provider.LOCAL
KNOWN_DOMAINS = frozenset(domain.value for domain in Domain)
_HEX = frozenset("0123456789abcdefABCDEF")
_UNRESERVED = "-._~"


def _component_value(value: Union[Enum, str, None]) -> str:
    if isinstance(value, Enum):
        return str(value.value)
    return "" if value is None else str(value)


def _decode_entity_id(raw: str) -> str:
    if not isinstance(raw, str) or not raw:
        raise TypedRefParseError("empty id")
    index = 0
    while index < len(raw):
        if raw[index] == "%":
            if (
                index + 2 >= len(raw)
                or raw[index + 1] not in _HEX
                or raw[index + 2] not in _HEX
            ):
                raise TypedRefParseError("malformed percent encoding in id")
            index += 3
        else:
            index += 1
    try:
        decoded = unquote_to_bytes(raw).decode("utf-8")
    except UnicodeDecodeError as exc:
        raise TypedRefParseError("invalid UTF-8 percent encoding in id") from exc
    if not decoded:
        raise TypedRefParseError("empty id")
    return decoded


def _encode_entity_id(value: str) -> str:
    if not isinstance(value, str) or not value:
        raise TypedRefParseError("empty id")
    return quote(value, safe=_UNRESERVED)


@dataclass(frozen=True)
class TypedRef:
    """Typed reference using the canonical four-segment v2 format."""

    domain: Union[Domain, str]
    type: Union[EntityType, str]
    provider: Union[Provider, str]
    id: str

    def __str__(self) -> str:
        domain_val = _component_value(self.domain).lower()
        type_val = _component_value(self.type).lower()
        provider_val = _component_value(self.provider).lower()
        return f"{domain_val}:{type_val}:{provider_val}:{_encode_entity_id(self.id)}"

    @property
    def entity_type(self) -> str:
        return _component_value(self.type)

    @property
    def entity_id(self) -> str:
        return self.id

    def ref(self) -> str:
        return str(self)

    def canonical(self) -> str:
        return str(self)

    def is_zero(self) -> bool:
        return not self.domain and not self.type and not self.provider and not self.id

    def is_valid(self) -> bool:
        try:
            domain = _component_value(self.domain).lower()
            _validate_domain(domain)
            _validate_entity_type(domain, _component_value(self.type))
            _validate_provider(_component_value(self.provider))
            _validate_id(self.id)
        except TypedRefParseError:
            return False
        return True


class TypedRefParseError(ValueError):
    """Error raised when a typed_ref is malformed."""


def parse_typed_ref(s: str) -> TypedRef:
    """Parse a strict v2 four-segment typed_ref."""

    if not isinstance(s, str):
        raise TypedRefParseError("ref must be a string")
    s = s.strip()
    if not s:
        raise TypedRefParseError("empty ref")
    parts = s.split(":")
    if len(parts) != 4:
        raise TypedRefParseError(
            f"invalid ref format: {s} (v2 requires <domain>:<type>:<provider>:<id>)"
        )
    return _parse_four_segment(parts)


def _parse_four_segment(parts: list[str]) -> TypedRef:
    domain_str = parts[0].lower()
    _validate_domain(domain_str)
    entity_type_str = parts[1].lower()
    _validate_entity_type(domain_str, entity_type_str)
    provider_str = parts[2].lower()
    _validate_provider(provider_str)
    id_part = _decode_entity_id(parts[3])
    return TypedRef(
        domain=Domain(domain_str),
        type=_known_entity_type_or_string(entity_type_str),
        provider=_known_provider_or_string(provider_str),
        id=id_part,
    )


def _validate_domain(domain: str) -> None:
    if domain not in KNOWN_DOMAINS:
        raise TypedRefParseError(
            f"invalid domain: {domain} (expected memx, agent-taskstate, or tracker)"
        )
    if "%" in domain:
        raise TypedRefParseError("percent encoding is only allowed in id")


def _validate_entity_type(domain: str, entity_type: str) -> None:
    if not entity_type:
        raise TypedRefParseError("empty entity type")
    if "%" in entity_type:
        raise TypedRefParseError("percent encoding is only allowed in id")


def _validate_provider(provider: str) -> None:
    if not provider:
        raise TypedRefParseError("empty provider")
    if "%" in provider:
        raise TypedRefParseError("percent encoding is only allowed in id")


def _validate_id(entity_id: str) -> None:
    _decode_entity_id(entity_id)


def _known_entity_type_or_string(value: str) -> Union[EntityType, str]:
    try:
        return EntityType(value)
    except ValueError:
        return value


def _known_provider_or_string(value: str) -> Union[Provider, str]:
    try:
        return Provider(value)
    except ValueError:
        return value


def new_typed_ref(entity_type: Union[EntityType, str], id: str) -> TypedRef:
    return TypedRef(
        domain=Domain.MEMX,
        type=entity_type,
        provider=DEFAULT_PROVIDER,
        id=id,
    )


def new_typed_ref_with_provider(
    domain: Union[Domain, str],
    entity_type: Union[EntityType, str],
    provider: Union[Provider, str],
    id: str,
) -> TypedRef:
    return TypedRef(
        domain=domain,
        type=entity_type,
        provider=provider,
        id=id,
    )


def must_parse_typed_ref(s: str) -> TypedRef:
    return parse_typed_ref(s)

File: api/typed_ref.go
package api

import (
	"fmt"
	"net/url"
	"strings"
	"unicode/utf8"
)

// Domain は typed_ref のシステム領域を表す。
type Domain string

const (
	DomainMemx           Domain = "memx"
	DomainAgentTaskstate Domain = "agent-taskstate"
	DomainTracker        Domain = "tracker"
)

// EntityType はエンティティ種別。未知の値も拡張用に受理する。
type EntityType string

const (
	EntityTypeEvidence      EntityType = "evidence"
	EntityTypeEvidenceChunk EntityType = "evidence_chunk"
	EntityTypeKnowledge     EntityType = "knowledge"
	EntityTypeArtifact      EntityType = "artifact"
	EntityTypeLineage       EntityType = "lineage"
)

// Provider はデータソース。未知の値も拡張用に受理する。
type Provider string

const (
	ProviderLocal  Provider = "local"
	ProviderJira   Provider = "jira"
	ProviderGitHub Provider = "github"
	ProviderLinear Provider = "linear"
)

const DefaultProvider Provider = ProviderLocal

// TypedRef はv2の4セグメント型付き参照。
type TypedRef struct {
	Domain   Domain     `json:"domain"`
	Type     EntityType `json:"type"`
	Provider Provider   `json:"provider"`
	ID       string     `json:"id"`
}

// String はdomain/type/providerを小文字化し、IDをRFC 3986形式で符号化する。
func (r TypedRef) String() string {
	return fmt.Sprintf(
		"%s:%s:%s:%s",
		strings.ToLower(string(r.Domain)),
		strings.ToLower(string(r.Type)),
		strings.ToLower(string(r.Provider)),
		percentEncodeID(r.ID),
	)
}

func (r TypedRef) MarshalText() ([]byte, error) {
	return []byte(r.String()), nil
}

func (r *TypedRef) UnmarshalText(text []byte) error {
	ref, err := ParseTypedRef(string(text))
	if err != nil {
		return err
	}
	*r = ref
	return nil
}

// ParseTypedRef はv2の4セグメント形式だけを受理する。
func ParseTypedRef(s string) (TypedRef, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return TypedRef{}, fmt.Errorf("empty ref")
	}
	parts := strings.Split(s, ":")
	if len(parts) != 4 {
		return TypedRef{}, fmt.Errorf("invalid ref format: %s (v2 requires <domain>:<type>:<provider>:<id>)", s)
	}
	return parseFourSegment(parts)
}

func parseFourSegment(parts []string) (TypedRef, error) {
	domain := Domain(strings.ToLower(parts[0]))
	if err := validateDomain(domain); err != nil {
		return TypedRef{}, err
	}
	entityType := EntityType(strings.ToLower(parts[1]))
	if err := validateEntityType(domain, entityType); err != nil {
		return TypedRef{}, err
	}
	provider := Provider(strings.ToLower(parts[2]))
	if err := validateProvider(provider); err != nil {
		return TypedRef{}, err
	}
	id, err := decodeEntityID(parts[3])
	if err != nil {
		return TypedRef{}, err
	}
	return TypedRef{Domain: domain, Type: entityType, Provider: provider, ID: id}, nil
}

func validateDomain(domain Domain) error {
	switch domain {
	case DomainMemx, DomainAgentTaskstate, DomainTracker:
		return nil
	default:
		return fmt.Errorf("invalid domain: %s (expected memx, agent-taskstate, or tracker)", domain)
	}
}

func validateEntityType(_ Domain, entityType EntityType) error {
	if entityType == "" {
		return fmt.Errorf("empty entity type")
	}
	if strings.Contains(string(entityType), "%") {
		return fmt.Errorf("percent encoding is only allowed in id")
	}
	return nil
}

func validateProvider(provider Provider) error {
	if provider == "" {
		return fmt.Errorf("empty provider")
	}
	if strings.Contains(string(provider), "%") {
		return fmt.Errorf("percent encoding is only allowed in id")
	}
	return nil
}

func decodeEntityID(raw string) (string, error) {
	if raw == "" {
		return "", fmt.Errorf("empty id in ref")
	}
	for i := 0; i < len(raw); i++ {
		if raw[i] != '%' {
			continue
		}
		if i+2 >= len(raw) || !isHex(raw[i+1]) || !isHex(raw[i+2]) {
			return "", fmt.Errorf("malformed percent encoding in id")
		}
		i += 2
	}
	decoded, err := url.PathUnescape(raw)
	if err != nil {
		return "", fmt.Errorf("malformed percent encoding in id: %w", err)
	}
	if decoded == "" {
		return "", fmt.Errorf("empty id in ref")
	}
	if !utf8.ValidString(decoded) {
		return "", fmt.Errorf("invalid UTF-8 percent encoding in id")
	}
	return decoded, nil
}

func percentEncodeID(id string) string {
	var b strings.Builder
	const hex = "0123456789ABCDEF"
	for i := 0; i < len(id); i++ {
		c := id[i]
		if isUnreserved(c) {
			b.WriteByte(c)
			continue
		}
		b.WriteByte('%')
		b.WriteByte(hex[c>>4])
		b.WriteByte(hex[c&0x0f])
	}
	return b.String()
}

func isUnreserved(c byte) bool {
	return (c >= 'a' && c <= 'z') ||
		(c >= 'A' && c <= 'Z') ||
		(c >= '0' && c <= '9') ||
		c == '-' || c == '.' || c == '_' || c == '~'
}

func isHex(c byte) bool {
	return (c >= '0' && c <= '9') ||
		(c >= 'a' && c <= 'f') ||
		(c >= 'A' && c <= 'F')
}

func MustParseTypedRef(s string) TypedRef {
	ref, err := ParseTypedRef(s)
	if err != nil {
		panic(err)
	}
	return ref
}

func NewTypedRef(entityType EntityType, id string) TypedRef {
	return TypedRef{Domain: DomainMemx, Type: entityType, Provider: DefaultProvider, ID: id}
}

func NewTypedRefWithProvider(domain Domain, entityType EntityType, provider Provider, id string) TypedRef {
	return TypedRef{Domain: domain, Type: entityType, Provider: provider, ID: id}
}

func (r TypedRef) Ref() string {
	return r.String()
}

func (r TypedRef) IsZero() bool {
	return r.Domain == "" && r.Type == "" && r.Provider == "" && r.ID == ""
}

func (r TypedRef) IsValid() bool {
	if validateDomain(Domain(strings.ToLower(string(r.Domain)))) != nil {
		return false
	}
	if validateEntityType(r.Domain, EntityType(strings.ToLower(string(r.Type)))) != nil {
		return false
	}
	if validateProvider(Provider(strings.ToLower(string(r.Provider)))) != nil {
		return false
	}
	_, err := decodeEntityID(r.ID)
	return err == nil
}

func (r TypedRef) Canonical() string {
	return r.String()
}
